design/api.md

```diff
@@ -0,0 +1,193 @@
+
+![Alt text](architecture.excalidraw.svg)
+
+## @fission/auth
+
+### API
+
+```ts
+export type AuthResponse<R> =
+  | {
+      error: undefined
+      result: R
+    }
+  | {
+      error: Error
+      result: undefined
+    }
+
+interface RegisterCredentials {
+  username: string
+  email: string
+  options: {
+    emailRedirectUrl?: string
+  }
+}
+interface Credentials {
+  username: string
+}
+
+interface Account {
+  did: string
+  username: string
+  email: string
+}
+
+interface Session {
+  agent: Agent
+  account: Account
+  expiresAt: number
+  identifierDelegation: Ucan
+  accountDelegation: Ucan
+}
+
+interface AuthConfig {
+  name?: string // defaults to origin
+  debug?: boolean
+  identifier: IdentifierProvider
+  agent: Agent // Handles signatures, dids, ucans and storage
+  client: FissionClient // Handles server requests should be a generic client for a Auth UCAN Protocol
+}
+
+declare class Auth {
+  constructor(config: AuthConfig)
+}
+
+interface IAuth {
+  new (config: AuthConfig): Auth
+  /**
+   * Trigger an email flow to register a user
+   * Signs a Register UCAN with an Identifier
+   */
+  register(credentials: RegisterCredentials): Promise<
+    AuthResponse<{
+      account: Account
+    }>
+  >
+
+  /**
+   * Used to exchange a code from an email link for a session
+   */
+  exchangeCodeForSession(code: string): Promise<
+    AuthResponse<{
+      session: Session
+      account: Account
+    }>
+  >
+
+  /**
+   * Requests an Identifier Delegation and IF NEEDED Server Delegations
+   *
+   * Should we control Session TTL with Identifier Delegation exp and/or Server Delegation exp?
+   */
+  login(credentials: Credentials): Promise<
+    AuthResponse<{
+      session: Session
+      account: Account
+    }>
+  >
+
+  /**
+   * Delete Identifier and server Delegations
+   *
+   * Should we delete all agent data as well?
+   */
+  logout(session: Session): Promise<void>
+
+  /**
+   * Subscribe to auth state changes
+   */
+  onStateChange(
+    cb: (
+      state: 'Login' | 'Logout' | 'Register' | 'Expired',
+      session: null | Session
+    ) => void
+  ): UnsubscribeFn
+}
+```
+
+### Notes
+
+- Should keep tabs in sync when in browser
+- Revocation api ?
+
+### Identifier Providers
+
+Providers SHOULD:
+
+- Depend only on UCAN (did:key, signer) and optionally on a [channel](#channels). They can have other dependencies but they should be environment specific. They SHOULD NOT depend on an [Agent](#agent).
+- Be able to receive a DID and a set of capabilities and return a powerbox style UCAN with those capabilities.
+- Be able to send a root delegation over a [channel](#channels) to another device or origin. (see [Delegated](#types))
+
+A powerbox style delegation to an Agent should be treated as a Session.
+
+#### Types
+
+- Local: Only exists on device normally sandboxed to an origin (webcrypto, wallet, metamask snap and local passkeys)
+- Synced: Exists on device and is synced with a server (synced passkey)
```

Comment:
Where a synced passkey is managed by the browser and has the same DID across devices, or could it run into the problem where there's multiple DIDs for the same user on the same browser that you alluded to the other day?

Reply:
Yep makes sense. Registering a new passkey is possible two ways, and you're right that we should have APIs the email one:

- Device linking: use one of your client-side UCANs to delegate permissions
- Email recovery: we send you a code, and you tell us which DID to delegate to (similar to Bluesky and other systems, could be extended with OAuth logins etc down the line, though bleh rooting security in OAuth)

Reply:
💯

Reply:
something like metamask could also be considered Synced if we assume the User manually syncs the mnemonic, i just dont think we should... That why metamask is local in the text, and we should either do Local/Delegated or Local/Remote linking.